presto-main/src/main/java/com/facebook/presto/util/MathUtils.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.util;
+
+import java.util.Random;
+
+public final class MathUtils
+{
+    private static final int NUMBER_OF_ESTIMATES = 50;
+    private static final int NUMBER_OF_PROBES = 100;
+    private static final int NUMBER_OF_VALUES = 500;
+    private static final double[] COLLISION_ESTIMATES = new double[NUMBER_OF_ESTIMATES + 1];
+
+    static {
+        generateEstimates();
+    }
+
+    public static double estimateNumberOfHashCollisions(int numberOfValues, int hashSize)
+    {
+        if (hashSize == 0) {
+            return 0d;
+        }
+
+        return COLLISION_ESTIMATES[NUMBER_OF_ESTIMATES * numberOfValues / hashSize] * numberOfValues / NUMBER_OF_VALUES;
```

Comment:
You could easily improve estimation quality by extrapolation of two closest values - it's extremely simple:

```
double lowerEstimation = COLLISION_ESTIMATES[floor(NUMBER_OF_ESTIMATES * numberOfValues / hashSize)];
double upperEstimation = COLLISION_ESTIMATES[ceil(NUMBER_OF_ESTIMATES * numberOfValues / hashSize)];
double estimationDistanceFromLower = NUMBER_OF_ESTIMATES * numberOfValues / hashSize - floor(NUMBER_OF_ESTIMATES * numberOfValues / hashSize);
double estimation = lowerEstimation * estimationDistanceFromLower + upperEstimation * (1 - estimationDistanceFromLower)
return estimation * numberOfValues / NUMBER_OF_VALUES;
```